plat/rockchip/rk3399/drivers/pmu/pmu.c

```diff
@@ -67,6 +67,350 @@ __attribute__ ((section("tzfw_coherent_mem")))
 #endif
 ;/* coheront */
 
+static void pmu_bus_idle_req(uint32_t bus, uint32_t idle)
+{
+	uint32_t bus_id = BIT(bus);
+	uint32_t bus_req;
+	uint32_t wait_cnt = 0;
+	uint32_t bus_state, bus_ack;
+
+	if (idle)
+		bus_req = BIT(bus);
+	else
+		bus_req = idle;
+
+	bus_state = bus_req;
+	bus_ack = bus_req;
+
+	mmio_clrsetbits_32(PMU_BASE + PMU_BUS_IDLE_REQ, bus_id, bus_req);
+
+	while (((mmio_read_32(PMU_BASE + PMU_BUS_IDLE_ST) & bus_id)
+	       != bus_state) &&  ((mmio_read_32(PMU_BASE + PMU_BUS_IDLE_ACK) &
+	       bus_id) != bus_ack)) {
+		wait_cnt++;
+		if (!(wait_cnt % 1000)) {
+			INFO("%s:st=%x(%x)\n", __func__,
+			     mmio_read_32(PMU_BASE + PMU_BUS_IDLE_ST),
+			     bus_state);
+			INFO("%s:st=%x(%x)\n", __func__,
+			     mmio_read_32(PMU_BASE + PMU_BUS_IDLE_ACK),
+			     bus_ack);
+		}
+	}
```

Comment:
Potentially could hang here. Consider breaking out of loop after a certain time and cancelling op.